staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,621 @@
+	fq := &queueSetImpl{
+		wg:               wg,
+		name:             name,
+		queues:           initQueues(desiredNumQueues),
+		clk:              clk,
+		vt:               0,
+		lastRealTime:     clk.Now(),
+		desiredNumQueues: desiredNumQueues,
+		concurrencyLimit: concurrencyLimit,
+		queueLengthLimit: queueLengthLimit,
+		requestWaitLimit: requestWaitLimit,
+	}
+	return fq
+}
+
+// LockAndSyncTime is used to ensure that the virtual time of a queueSetImpl
+// is synced everytime its fields are accessed
+func (qs *queueSetImpl) LockAndSyncTime() {
+	qs.lock.Lock()
+	qs.synctime()
+}
+
+// SetConfiguration is used to set the configuration for a queueSetImpl
+// update handling for when fields are updated is handled here as well -
+// eg: if desiredNumQueues is increased, SetConfiguration reconciles by
+// adding more queues.
+func (qs *queueSetImpl) SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit int, requestWaitLimit time.Duration) {
+	// TODO(aaron-prindle) verify updating queues makes sense here vs elsewhere
+
+	// lock required as method can change Queues which has its indexes and length used
+	// concurrently
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+
+	// Adding queues is the only thing that requires immediate action
+	// Removing queues is handled by omitting indexes >desiredNumQueues from
+	// chooseQueueIdx
+
+	numQueues := len(qs.queues)
+	if desiredNumQueues > numQueues {
+		qs.addQueues(desiredNumQueues - numQueues)
+	}
+
+	qs.concurrencyLimit = concurrencyLimit
+	qs.desiredNumQueues = desiredNumQueues
+	qs.queueLengthLimit = queueLengthLimit
+	qs.requestWaitLimit = requestWaitLimit
+}
+
+// TimeoutOldRequestsAndRejectOrEnqueue encapsulates the lock sharing logic required
+// to validated and enqueue a request for the queueSetImpl/QueueSetSystem:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a packet and enqueue
+// returns true on a successful enqueue
+// returns false in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
+func (qs *queueSetImpl) TimeoutOldRequestsAndRejectOrEnqueue(hashValue uint64, handSize int32) *Request {
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.ChooseQueueIdx(hashValue, int(handSize))
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a packet and enqueue
+	pkt := &Request{
```

Comment:
The conceptual basis of this is some work in the networking domain, so the initial implementation was written using the analogous networking terms.  I agree that this work should use "request" rather than "packet".

Reply:
fixed in https://github.com/kubernetes/kubernetes/pull/81834/commits/480ab702c5df71a0fa95a05d40de7baebf2f3d6a